### PR TITLE
choose_weighted distribution tuple

### PR DESCRIPTION
**Pull request by justinsalamon**
Add support for a `choose_weighted` distribution tuple of the form `("choose_weighted", listA, listB)` where `listA` and `listB` are of the same length, A contains the items to select from and B the corresponding probabilities. `listB` must contain a valid set of probabilities in the range [0, 1] that sum to 1. Addresses #143.

### CODE REVIEW

**Comment by justinsalamon** [comment]
@pseeth ready for a first CR!!

File: scaper/core.py [hunk 1]
```diff
@@ -450,6 +468,27 @@ def _validate_distribution(dist_tuple):
...
+        if not isinstance(dist_tuple[1], list) or \
+           not isinstance(dist_tuple[2], list) or \
+           len(dist_tuple[1]) != len(dist_tuple[2]):
+           msg = ('The 2nd and 3rd items of the "choose_weighted" distribution tuple '
+                  'must be lists of the same length.')
+           raise ScaperError(msg)
+        probabilities = np.asarray(dist_tuple[2])
+        if probabilities.min() < 0 or probabilities.max() > 1:
+            msg = ('Values in the probabilities list of the "choose_weighted" '
+                   'distirbution tuple must be in the range [0, 1].')
```
Comment (pseeth) [inline scaper/core.py:486]: distirbution -> distribution

File: scaper/core.py [hunk 2]
```diff
@@ -1355,6 +1416,7 @@ def _instantiate_event(self, event, isbackground=False,
             allowed_labels = self.fg_labels
 
         # determine label
+        # special case: choose tuple with empty list
         if event.label[0] == "choose" and not event.label[1]:
```
Comment (pseeth) [inline scaper/core.py:1420]: Why isn't this also happening for `choose_weighted` here?

File: scaper/core.py [hunk 3]
```diff
@@ -1380,9 +1442,9 @@ def _instantiate_event(self, event, isbackground=False,
             used_labels.append(label)
 
         # determine source file
+        # special case: choose tuple with empty list
         if event.source_file[0] == "choose" and not event.source_file[1]:
```
Comment (pseeth) [inline scaper/core.py:1446]: Why isn't this also happening for `choose_weighted`?

Comment (pseeth) [reply scaper/core.py:1446]: Oh so the lists of labels and probabilities have to match in length. Got it.

**Review by pseeth — APPROVED** [review]
I think it's good enough for a release candidate. Can merge, but let's release it as `...rc0` for now and have it stress tested before bumping to a proper release. I think we can still merge to master and then cut a release though.

I put some minor comments but lgtm.

File: scaper/core.py (hunk 1 above)
Comment (justinsalamon) [reply scaper/core.py:486]: fixed

File: scaper/core.py (hunk 2 above)
Comment (justinsalamon) [reply scaper/core.py:1420]: `choose` has a unique case when you submit `("choose", [])`, where we need to populate the list from files (or folders) on disk, i.e. we need to modify the distribution tuple before sampling from it.

In all other cases, including `choose_weighted`, there's no need to change the tuple before sampling.

File: scaper/core.py (hunk 3 above)
Comment (justinsalamon) [reply scaper/core.py:1446]: same as last comment, it's a special case where we need to update the "choose" tuple with info from disk

**Pull request merged**